tests/network/tincans.rs

```diff
@@ -0,0 +1,162 @@
+        right: TinCan {
+            send: Box::new(RecordCloseSink::new(
+                PollSender::new(right_send).sink_map_err(|e| {
+                    NetworkError::Error(format!(
+                        "unexpected failure to send on blocking channel: {:?}",
+                        e
+                    ))
+                }),
+                right_sink_closed.clone(),
+            )),
+            recv: Box::new(ReceiverStream::new(right_recv).map(Ok)),
+            sink_closed: right_sink_closed,
+        },
+    }
+}
+
+/// Create a single tincan which sends messages to nowhere and receives nothing.
+pub(crate) fn tincan_to_nowhere() -> TinCan {
+    let sink_closed = Arc::new(AtomicBool::new(false));
+    TinCan {
+        send: Box::new(RecordCloseSink::new(
+            futures::sink::drain().sink_map_err(|_e| unreachable!()),
+            sink_closed.clone(),
+        )),
+        recv: Box::new(futures::stream::pending()),
+        sink_closed,
+    }
+}
+
+pub(crate) fn connect_repos(left: &RepoHandle, right: &RepoHandle) {
+    let TinCans {
+        left:
+            TinCan {
+                send: left_send,
+                recv: left_recv,
+                ..
+            },
+        right:
+            TinCan {
+                send: right_send,
+                recv: right_recv,
+                ..
+            },
+    } = tincans();
+    left.new_remote_repo(right.get_repo_id().clone(), left_recv, left_send);
+    right.new_remote_repo(left.get_repo_id().clone(), right_recv, right_send);
+}
+
+/// A wrapper around a `Sink` which records whether `poll_close` has ever been called
+struct RecordCloseSink<S> {
+    inner: S,
+    closed: Arc<AtomicBool>,
+}
+
+impl<S> RecordCloseSink<S> {
+    fn new(inner: S, closed: Arc<AtomicBool>) -> Self {
+        Self { inner, closed }
+    }
+}
+
+impl<S> Sink<RepoMessage> for RecordCloseSink<S>
+where
+    S: Sink<RepoMessage, Error = NetworkError> + Unpin,
+{
+    type Error = NetworkError;
+    fn poll_ready(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        let inner = unsafe { std::pin::Pin::new_unchecked(&mut self.get_unchecked_mut().inner) };
+        inner.poll_ready(cx)
+    }
+    fn start_send(self: std::pin::Pin<&mut Self>, item: RepoMessage) -> Result<(), Self::Error> {
+        let inner = unsafe { std::pin::Pin::new_unchecked(&mut self.get_unchecked_mut().inner) };
+        inner.start_send(item)
+    }
+    fn poll_flush(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        let inner = unsafe { std::pin::Pin::new_unchecked(&mut self.get_unchecked_mut().inner) };
+        inner.poll_flush(cx)
+    }
+    fn poll_close(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.closed
```

Comment:
I would rather see this set to true only if inner returns ready, so that it covers the need to keep polling the sink if the first poll doesn't complete the close.

Reply:
Done